components/sup/src/manager/service/health.rs

```diff
@@ -26,12 +46,167 @@ impl From<i8> for HealthCheck {
+             };
+
+             trace!("Next health check for {} in {}", service_group, interval);
+
+             let instant =
+                 Instant::now().checked_add(interval.into())
+                               .expect("This should never happen with normal health check \
+                                        interval sizes");
+             tokio_timer::timer::Handle::current().delay(instant)
+                                                  .map_err(move |timer_error| {
+                                                      if timer_error.is_shutdown() {
+                                                          warn!("Timer for {} health check shut \
+                                                                 down!",
+                                                                service_group);
+                                                      }
+                                                      if timer_error.is_at_capacity() {
+                                                          warn!("Timer for {} health check is at \
+                                                                 capacity!",
+                                                                service_group);
+                                                      }
+                                                  })
+         })
+    }
+
+    /// Repeatedly runs a health check, followed by an appropriate
+    /// delay, forever.
+    pub fn check_repeatedly(self) -> impl Future<Item = (), Error = ()> {
+        future::loop_fn(self, move |state| {
+            // TODO (CM): If we wanted to keep track of how many times
+            // a health check has failed in the past X executions, or
+            // do similar historical tracking, here's where we'd do
+            // it.
+            let service_group = state.service_group.clone();
+            state.clone().single_iteration().then(move |res| {
+                                                match res {
+                                                    Ok(_) => {
+                                                        trace!("Health check future for {} \
+                                                                succeeded; continuing loop",
+                                                               service_group);
+                                                        Ok(Loop::Continue(state))
+                                                    }
+                                                    Err(_) => {
+                                                        trace!("Health check future for {} \
+                                                                failed failed; continuing loop.",
+                                                               service_group);
+                                                        Ok(Loop::Continue(state))
+                                                    }
+                                                }
```

Comment:
Since this is already pretty extremely indented and we're not doing anything with the *contents* of the `Result` variants, we could simplify:
```rust
                                                if res.is_ok() {
                                                    trace!("Health check future for {} \
                                                            succeeded; continuing loop",
                                                           service_group);
                                                    Ok(Loop::Continue(state))
                                                } else {
                                                    trace!("Health check future for {} failed \
                                                            failed; continuing loop.",
                                                           service_group);
                                                    Ok(Loop::Continue(state))
                                                }
```